core/src/commonMain/kotlin/com/powersync/attachments/sync/SyncingService.kt

```diff
@@ -0,0 +1,278 @@
+import com.powersync.attachments.AttachmentService
+import com.powersync.attachments.AttachmentState
+import com.powersync.attachments.LocalStorage
+import com.powersync.attachments.RemoteStorage
+import com.powersync.attachments.SyncErrorHandler
+import com.powersync.utils.throttle
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.IO
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.asSharedFlow
+import kotlinx.coroutines.flow.buffer
+import kotlinx.coroutines.flow.merge
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.seconds
+
+/**
+ * Service used to sync attachments between local and remote storage
+ */
+public class SyncingService(
+    private val remoteStorage: RemoteStorage,
+    private val localStorage: LocalStorage,
+    private val attachmentsService: AttachmentService,
+    private val getLocalUri: suspend (String) -> String,
+    private val errorHandler: SyncErrorHandler?,
+    private val logger: Logger,
+    private val syncThrottle: Duration = 5.seconds,
+) {
+    private val scope = CoroutineScope(Dispatchers.IO)
+    private val mutex = Mutex()
+    private var syncJob: Job? = null
+
+    /**
+     * Used to trigger the sync process either manually or periodically
+     */
+    private val syncTriggerFlow = MutableSharedFlow<Unit>(replay = 0)
+
+    /**
+     * Starts syncing operations
+     */
+    public suspend fun startSync(period: Duration = 30.seconds): Unit =
+        mutex.withLock {
+            syncJob?.cancel()
+            syncJob?.join()
+
+            syncJob =
+                scope.launch {
+                    val watchJob =
+                        launch {
+                            merge(
+                                // Handles manual triggers for sync events
+                                syncTriggerFlow.asSharedFlow(),
+                                // Triggers the sync process whenever an underlaying change to the
+                                // attachments table happens
+                                attachmentsService
+                                    .watchActiveAttachments(),
+                            )
+                                // We only use these flows to trigger the process. We can skip multiple invocations
+                                // while we are processing. We will always process on the trailing edge.
+                                // This buffer operation should automatically be applied to all merged sources.
+                                .buffer(1, onBufferOverflow = BufferOverflow.DROP_OLDEST)
+                                .throttle(syncThrottle.inWholeMilliseconds)
```

Comment:
Given that it's internal, I wonder if we should just change `throttle` to take a `Duration` 🤔

Reply:
Yeah. Changed this to a Duration now.